From 0221775227fbde88e9308166434d6d5f55fccc35 Mon Sep 17 00:00:00 2001
From: Henry Wang <Henry.Wang@arm.com>
Date: Sat, 8 Oct 2022 07:09:05 +0000
Subject: [PATCH] tools/xl-psr: Consolidate CAT hwinfo to a single helper

Currently, the `xl psr-hwinfo` command calls two different helpers
namely `psr_l2_cat_hwinfo()` and `psr_l3_cat_hwinfo()` for L2 and
L3 cache respectively, and the implementation of these two helpers
are mostly similar and partly duplicated.

As different types of cache would be introduced in the future and
it is not a good practice to further duplicate the implementation,
this commit introduces `enum psr_cat_type` to distinguish the CAT
type and consolidates the above-mentioned two helpers to a single
helper `psr_cat_hwinfo(enum psr_cat_type)`. Therefore new types of
cache can be easily added by extending the enum and the new helper.

No functional change is intended.

Signed-off-by: Henry Wang <Henry.Wang@arm.com>
Upstream-Status: Pending
Issue-Id: SCM-5244
Change-Id: I401186ecdccf42a87bd494505b82e3d4a312c4e6
---
 tools/xl/xl_psr.c | 78 ++++++++++++++++++++++++-----------------------
 1 file changed, 40 insertions(+), 38 deletions(-)

diff --git a/tools/xl/xl_psr.c b/tools/xl/xl_psr.c
index afdfff2c39..38750a149a 100644
--- a/tools/xl/xl_psr.c
+++ b/tools/xl/xl_psr.c
@@ -33,6 +33,11 @@
 #define CAPACITY_BITMASK_NAME "CPBM"
 #endif
 
+enum psr_cat_type {
+    PSR_CAT_L2 = 2,
+    PSR_CAT_L3,
+};
+
 static int psr_cmt_hwinfo(void)
 {
     int rc;
@@ -300,30 +305,52 @@ int main_psr_cmt_show(int argc, char **argv)
     return ret;
 }
 
-static int psr_l3_cat_hwinfo(void)
+static int psr_cat_hwinfo(enum psr_cat_type type)
 {
     int rc;
     unsigned int i, nr;
     uint32_t l3_cache_size;
     libxl_psr_cat_info *info;
+    char *cache_type;
 
-    rc = libxl_psr_cat_get_info(ctx, &info, &nr, 3);
+    switch (type)
+    {
+    case PSR_CAT_L2:
+        cache_type = " L2";
+        break;
+    case PSR_CAT_L3:
+    default:
+        cache_type = "";
+        break;
+    }
+
+    rc = libxl_psr_cat_get_info(ctx, &info, &nr, type);
     if (rc)
         return rc;
 
-    printf("Cache Allocation Technology (CAT):\n");
+    printf("Cache Allocation Technology (CAT):%s\n", cache_type);
 
     for (i = 0; i < nr; i++) {
-        rc = libxl_psr_cmt_get_l3_cache_size(ctx, info[i].id, &l3_cache_size);
-        if (rc) {
-            fprintf(stderr, "Failed to get l3 cache size for socket:%d\n",
-                    info[i].id);
-            goto out;
+        if (type == PSR_CAT_L3)
+        {
+            rc = libxl_psr_cmt_get_l3_cache_size(ctx, info[i].id,
+                                                 &l3_cache_size);
+            if (rc) {
+                fprintf(stderr, "Failed to get l3 cache size for socket:%d\n",
+                        info[i].id);
+                goto out;
+            }
         }
+
         printf("%-16s: %u\n", "Socket ID", info[i].id);
-        printf("%-16s: %uKB\n", "L3 Cache", l3_cache_size);
-        printf("%-16s: %s\n", "CDP Status",
-               info[i].cdp_enabled ? "Enabled" : "Disabled");
+
+        if (type == PSR_CAT_L3)
+        {
+            printf("%-16s: %uKB\n", "L3 Cache", l3_cache_size);
+            printf("%-16s: %s\n", "CDP Status",
+                   info[i].cdp_enabled ? "Enabled" : "Disabled");
+        }
+
         printf("%-16s: %u\n", "Maximum "PARTITION_IDENTIFIER_NAME,
                info[i].cos_max);
         printf("%-16s: %u\n", CAPACITY_BITMASK_NAME" length", info[i].cbm_len);
@@ -517,31 +544,6 @@ out:
     return rc;
 }
 
-static int psr_l2_cat_hwinfo(void)
-{
-    int rc;
-    unsigned int i, nr;
-    libxl_psr_cat_info *info;
-
-    rc = libxl_psr_cat_get_info(ctx, &info, &nr, 2);
-    if (rc)
-        return rc;
-
-    printf("Cache Allocation Technology (CAT): L2\n");
-
-    for (i = 0; i < nr; i++) {
-        /* There is no CMT on L2 cache so far. */
-        printf("%-16s: %u\n", "Socket ID", info[i].id);
-        printf("%-16s: %u\n", "Maximum COS", info[i].cos_max);
-        printf("%-16s: %u\n", "CBM length", info[i].cbm_len);
-        printf("%-16s: %#llx\n", "Default CBM",
-               (1ull << info[i].cbm_len) - 1);
-    }
-
-    libxl_psr_cat_info_list_free(info, nr);
-    return rc;
-}
-
 int main_psr_mba_show(int argc, char **argv)
 {
     int opt;
@@ -786,11 +788,11 @@ int main_psr_hwinfo(int argc, char **argv)
         ret = psr_cmt_hwinfo();
 
     if (!ret && (all || cat))
-        ret = psr_l3_cat_hwinfo();
+        ret = psr_cat_hwinfo(PSR_CAT_L3);
 
     /* L2 CAT is independent of CMT and L3 CAT */
     if (all || cat)
-        ret = psr_l2_cat_hwinfo();
+        ret = psr_cat_hwinfo(PSR_CAT_L2);
 
     /* MBA is independent of CMT and CAT */
     if (all || mba)
-- 
2.25.1

